// typescript-to-lua/dist/transformation/utils/language-extensions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getIterableExtensionKindForNode = exports.getIterableExtensionTypeForType = exports.isLuaIterable = exports.IterableExtensionKind = exports.getExtensionKindForSymbol = exports.getExtensionKindForNode = exports.getExtensionKindForType = exports.ExtensionKind = void 0;
const ts = require("typescript");
var ExtensionKind;
(function (ExtensionKind) {
    ExtensionKind["MultiFunction"] = "MultiFunction";
    ExtensionKind["RangeFunction"] = "RangeFunction";
    ExtensionKind["VarargConstant"] = "VarargConstant";
    ExtensionKind["AdditionOperatorType"] = "Addition";
    ExtensionKind["AdditionOperatorMethodType"] = "AdditionMethod";
    ExtensionKind["SubtractionOperatorType"] = "Subtraction";
    ExtensionKind["SubtractionOperatorMethodType"] = "SubtractionMethod";
    ExtensionKind["MultiplicationOperatorType"] = "Multiplication";
    ExtensionKind["MultiplicationOperatorMethodType"] = "MultiplicationMethod";
    ExtensionKind["DivisionOperatorType"] = "Division";
    ExtensionKind["DivisionOperatorMethodType"] = "DivisionMethod";
    ExtensionKind["ModuloOperatorType"] = "Modulo";
    ExtensionKind["ModuloOperatorMethodType"] = "ModuloMethod";
    ExtensionKind["PowerOperatorType"] = "Power";
    ExtensionKind["PowerOperatorMethodType"] = "PowerMethod";
    ExtensionKind["FloorDivisionOperatorType"] = "FloorDivision";
    ExtensionKind["FloorDivisionOperatorMethodType"] = "FloorDivisionMethod";
    ExtensionKind["BitwiseAndOperatorType"] = "BitwiseAnd";
    ExtensionKind["BitwiseAndOperatorMethodType"] = "BitwiseAndMethod";
    ExtensionKind["BitwiseOrOperatorType"] = "BitwiseOr";
    ExtensionKind["BitwiseOrOperatorMethodType"] = "BitwiseOrMethod";
    ExtensionKind["BitwiseExclusiveOrOperatorType"] = "BitwiseExclusiveOr";
    ExtensionKind["BitwiseExclusiveOrOperatorMethodType"] = "BitwiseExclusiveOrMethod";
    ExtensionKind["BitwiseLeftShiftOperatorType"] = "BitwiseLeftShift";
    ExtensionKind["BitwiseLeftShiftOperatorMethodType"] = "BitwiseLeftShiftMethod";
    ExtensionKind["BitwiseRightShiftOperatorType"] = "BitwiseRightShift";
    ExtensionKind["BitwiseRightShiftOperatorMethodType"] = "BitwiseRightShiftMethod";
    ExtensionKind["ConcatOperatorType"] = "Concat";
    ExtensionKind["ConcatOperatorMethodType"] = "ConcatMethod";
    ExtensionKind["LessThanOperatorType"] = "LessThan";
    ExtensionKind["LessThanOperatorMethodType"] = "LessThanMethod";
    ExtensionKind["GreaterThanOperatorType"] = "GreaterThan";
    ExtensionKind["GreaterThanOperatorMethodType"] = "GreaterThanMethod";
    ExtensionKind["NegationOperatorType"] = "Negation";
    ExtensionKind["NegationOperatorMethodType"] = "NegationMethod";
    ExtensionKind["BitwiseNotOperatorType"] = "BitwiseNot";
    ExtensionKind["BitwiseNotOperatorMethodType"] = "BitwiseNotMethod";
    ExtensionKind["LengthOperatorType"] = "Length";
    ExtensionKind["LengthOperatorMethodType"] = "LengthMethod";
    ExtensionKind["TableNewType"] = "TableNew";
    ExtensionKind["TableDeleteType"] = "TableDelete";
    ExtensionKind["TableDeleteMethodType"] = "TableDeleteMethod";
    ExtensionKind["TableGetType"] = "TableGet";
    ExtensionKind["TableGetMethodType"] = "TableGetMethod";
    ExtensionKind["TableHasType"] = "TableHas";
    ExtensionKind["TableHasMethodType"] = "TableHasMethod";
    ExtensionKind["TableSetType"] = "TableSet";
    ExtensionKind["TableSetMethodType"] = "TableSetMethod";
    ExtensionKind["TableAddKeyType"] = "TableAddKey";
    ExtensionKind["TableAddKeyMethodType"] = "TableAddKeyMethod";
})(ExtensionKind = exports.ExtensionKind || (exports.ExtensionKind = {}));
const extensionValues = new Set(Object.values(ExtensionKind));
function getExtensionKindForType(context, type) {
    const value = getPropertyValue(context, type, "__tstlExtension");
    if (value && extensionValues.has(value)) {
        return value;
    }
}
exports.getExtensionKindForType = getExtensionKindForType;
const excludedTypeFlags = ((1 << 18) - 1) | // All flags from Any...Never
    ts.TypeFlags.Index |
    ts.TypeFlags.NonPrimitive;
function getPropertyValue(context, type, propertyName) {
    if (type.flags & excludedTypeFlags)
        return;
    const property = type.getProperty(propertyName);
    if (!property)
        return undefined;
    const propertyType = context.checker.getTypeOfSymbolAtLocation(property, context.sourceFile);
    if (propertyType.isStringLiteral())
        return propertyType.value;
}
function getExtensionKindForNode(context, node) {
    const originalNode = ts.getOriginalNode(node);
    let type = context.checker.getTypeAtLocation(originalNode);
    if (ts.isOptionalChain(originalNode)) {
        type = context.checker.getNonNullableType(type);
    }
    return getExtensionKindForType(context, type);
}
exports.getExtensionKindForNode = getExtensionKindForNode;
function getExtensionKindForSymbol(context, symbol) {
    const type = context.checker.getTypeOfSymbolAtLocation(symbol, context.sourceFile);
    return getExtensionKindForType(context, type);
}
exports.getExtensionKindForSymbol = getExtensionKindForSymbol;
var IterableExtensionKind;
(function (IterableExtensionKind) {
    IterableExtensionKind["Iterable"] = "Iterable";
    IterableExtensionKind["Pairs"] = "Pairs";
    IterableExtensionKind["PairsKey"] = "PairsKey";
})(IterableExtensionKind = exports.IterableExtensionKind || (exports.IterableExtensionKind = {}));
function isLuaIterable(context, type) {
    return getPropertyValue(context, type, "__tstlIterable") !== undefined;
}
exports.isLuaIterable = isLuaIterable;
function getIterableExtensionTypeForType(context, type) {
    const value = getPropertyValue(context, type, "__tstlIterable");
    if (value && value in IterableExtensionKind) {
        return value;
    }
}
exports.getIterableExtensionTypeForType = getIterableExtensionTypeForType;
function getIterableExtensionKindForNode(context, node) {
    const type = context.checker.getTypeAtLocation(node);
    return getIterableExtensionTypeForType(context, type);
}
exports.getIterableExtensionKindForNode = getIterableExtensionKindForNode;
